### PR TITLE
[SM-911] add more validations for DAT group vin decoding

### PR DESCRIPTION
**Pull request by Asfiroth**
# Proposed Changes

### CODE REVIEW

File: internal/infrastructure/gateways/datgroup_api_service.go [hunk 1]
```diff
@@ -140,21 +140,37 @@ func parseXML(logger *zerolog.Logger, datgroupRespXML, vin string) (*DATGroupInf
...
 	} else {
 		response.Year = response.YearHigh
 	}
+
 	// series equipment
 	seriesEquipment := xmlquery.FindOne(vehicle, "//ns1:SeriesEquipment")
+
+	if seriesEquipment == nil {
+		logger.Err(err).Str("vin", vin).Msgf("failed to find series equipment node")
+		return response, nil
```
Comment (JamesReate) [inline internal/infrastructure/gateways/datgroup_api_service.go:164]: return nil, nil

File: internal/infrastructure/gateways/datgroup_api_service.go [hunk 2]
```diff
@@ -140,21 +140,37 @@ func parseXML(logger *zerolog.Logger, datgroupRespXML, vin string) (*DATGroupInf
...
+	}
+
 	seNodes := xmlquery.Find(seriesEquipment, "//ns1:EquipmentPosition")
+
+	if seNodes == nil {
+		logger.Err(err).Str("vin", vin).Msgf("failed to find series equipment nodes")
+		return response, nil
```
Comment (JamesReate) [inline internal/infrastructure/gateways/datgroup_api_service.go:171]: return nil, err

Comment (Asfiroth) [reply internal/infrastructure/gateways/datgroup_api_service.go:171]: so it should always be equipments?

File: internal/infrastructure/gateways/datgroup_api_service.go [hunk 3]
```diff
@@ -140,21 +140,37 @@ func parseXML(logger *zerolog.Logger, datgroupRespXML, vin string) (*DATGroupInf
...
 		response.Year = yr
 	} else {
 		response.Year = response.YearHigh
 	}
+
 	// series equipment
 	seriesEquipment := xmlquery.FindOne(vehicle, "//ns1:SeriesEquipment")
+
+	if seriesEquipment == nil {
+		logger.Err(err).Str("vin", vin).Msgf("failed to find series equipment node")
```
Comment (JamesReate) [inline internal/infrastructure/gateways/datgroup_api_service.go:163]: the err here would be from a different error, lets instantiate a new fmt.errorf(..) and then return that as well as log - although check the the caller already logs the error (may not be necessary to log just return nil, the error)

File: internal/infrastructure/gateways/datgroup_api_service.go [hunk 4]
```diff
@@ -164,9 +180,22 @@ func parseXML(logger *zerolog.Logger, datgroupRespXML, vin string) (*DATGroupInf
 		}
 		response.SeriesEquipment = append(response.SeriesEquipment, equipment)
 	}
+
 	// special equipment
 	specialEquipment := xmlquery.FindOne(vehicle, "//ns1:SpecialEquipment")
+
+	if specialEquipment == nil {
+		logger.Err(err).Str("vin", vin).Msgf("failed to find special equipment node")
```
Comment (JamesReate) [inline internal/infrastructure/gateways/datgroup_api_service.go:188]: same here

**Review by JamesReate — APPROVED** [review]

**Pull request merged**